### PR TITLE
99 update seed file with requests responses & 82 Restructure tests in describe blocks for each Route

### PR DESCRIPTION
**Pull request by radu-constantin**
Updated the seed file with a few requests.
  Bob's Neighborhood has 1 user and 1 request.
  Antonina's Neighborhood has 2 users and 2 requests.
  Shwetank's Neighborhood remains empty in case we need to test something against an empty neighborhood.
I've also separated the tests based on request method and route.

### CODE REVIEW

File: apps/backend/prisma/seed.ts
Comment (nina-dk) [inline apps/backend/prisma/seed.ts:1]: `getPasswordHash` and `connectUsertoNeighborhood` are duplicated

**Review by nina-dk — CHANGES_REQUESTED** [review]
Great work as always, Radu! I've left some comments all over 😅

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 1]
```diff
@@ -153,7 +90,7 @@ describe('When neighborhoods already exist in the db', () => {
     expect(response.body.error).toEqual('No Neighborhood found');
   });
 
-  test('GET /neighborhoods/id only returns the id, name, description and location of neighborhood if user is not logged in', async () => {
+  test('GET /neighborhoods/:id only returns the id, name, description and location of neighborhood if user is not logged in', async () => {
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:93]: Can you bring this up at our next meeting? I think this is the behavior we've decided on.

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 2]
```diff
@@ -36,94 +36,31 @@ beforeAll(async () => {
...
-
-    expect(deleteResponse.status).toEqual(401);
-    expect(numInitialNeighborhoods).toEqual(numCurrentNeighborhoods);
+describe('Tests for getting a single neighborhood: GET /neighborhoods/:id', () => {
+  beforeAll(async () => {
+    await seed();
   });
 
-  test('GET /neighborhoods/id existing id returns single neignborhood', async () => {
+  test('GET /neighborhoods/:id existing id returns single neignborhood', async () => {
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:63]: You really have an incredible attention for details 💯.
Will fix!

**Comment by radu-constantin** [comment]
Thanks for the review Antonina! Will implement the changes!

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 3]
```diff
@@ -153,7 +90,7 @@ describe('When neighborhoods already exist in the db', () => {
     expect(response.body.error).toEqual('No Neighborhood found');
   });
 
-  test('GET /neighborhoods/id only returns the id, name, description and location of neighborhood if user is not logged in', async () => {
+  test('GET /neighborhoods/:id only returns the id, name, description and location of neighborhood if user is not logged in', async () => {
     const neighborhood = await prismaClient.neighborhood.findFirst({
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:94]: I've added the type but I got a an error 
`Type '{ id: number; name: string; description: string | null; location: string | null; } | null' is not assignable to type 'NeighborhoodDetailsForNonMembers'.
  Type 'null' is not assignable to type 'NeighborhoodDetailsForNonMembers'.ts(2322)`
The solution I found was to add `as` NeighborhoodDetailsForNonMembers at the end.
Do you think it's ok to do that?

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 4]
```diff
@@ -199,19 +136,201 @@ describe('When neighborhoods already exist in the db', () => {
...
 
-  test('GET /neighborhoods return 200', async () => {
-    const response = await api.get('/api/neighborhoods');
-    expect(response.status).toEqual(200);
-    expect(response.body.length).toBe(0);
+  test('when user not logged in, unable to create neighborhood', async () => {
+    const initialNeighborhoods = await testHelpers.neighborhoodsInDb();
+    const numInitialNeighborhoods = initialNeighborhoods.length;
+
+    const postResponse = await api.post('/api/neighborhoods');
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts]: I've added the proper type, but can you explain why do you think this is necessary? I don't think that the `post` method can return anything other than a response.

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 5]
```diff
@@ -199,19 +136,201 @@ describe('When neighborhoods already exist in the db', () => {
   });
 });
 
-describe('When no neighborhood exists in the db', () => {
+describe('Tests for creating a single neighborhood: POST /neighborhoods/:id ', () => {
   beforeEach(async () => {
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:144]: Brilliant catch!

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 6]
```diff
@@ -199,19 +136,201 @@ describe('When neighborhoods already exist in the db', () => {
...
+  });
+});
+
+describe('Tests for deleting a single neighborhood: DELETE /neighborhoods/:id', () => {
+  beforeEach(async () => {
+    await seed();
+  });
+
+  test('DELETE /neighborhoods/:id removes a neighborhood', async () => {
+    const initialNeighborhoods = await testHelpers.neighborhoodsInDb();
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:255]: Done!

File: apps/backend/src/tests/neighborhood_api.test.ts
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:1]: I don't understand the first point, can you explain?

File: apps/backend/src/tests/seed.ts
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/seed.ts:1]: I think we can, but I'd like @MaxHass12 to also give his opinion on this

File: apps/backend/src/tests/neighborhood_api.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobertRodes) [reply apps/backend/src/tests/neighborhood_api.test.ts:94]: My understanding was that guest users would only see the all neighborhoods display, with a log in/sign up thing. My understanding was that the `GET /neighborhoods/:id` route was only accessible to logged-in users.

I'd like to suggest that we document features that we have decided on, and consider features that aren't documented as undecided. If we have a document that we all understand to be the "source of truth" for the features we're implementing, it will be easier to implement features.

File: apps/backend/src/tests/neighborhood_api.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (RobertRodes) [reply apps/backend/src/tests/neighborhood_api.test.ts]: I'm concerned that you're implementing a "phantom feature." What's the use case narrative for a guest user? (What are all the interactions that a guest user can have with the site?) Personally, I thought we had agreed that a guest user (a "not-logged-in" user) could only see the landing page, and that the landing page would contain a list of neighborhoods such as the rough one I put up last night, some stuff explaining what we're doing, and a Sign up/Sign in feature.

File: apps/backend/src/tests/neighborhood_api.test.ts
[the remark this answers is not in the document]
Comment (MaxHass12) [reply apps/backend/src/tests/neighborhood_api.test.ts:1]: Why not moving `await seed()` on `line 43` to the `beforeAll` block.

**Review by MaxHass12 — APPROVED** [review]
Looks all good. Well done.

File: apps/backend/src/tests/seed.ts
[the remark this answers is not in the document]
Comment (MaxHass12) [reply apps/backend/src/tests/seed.ts:1]: Test suite should be insulated from the app code. Hence test helpers, test seed should all be in separate files even though it is exactly the same code.

I am not super comfortable with using login routes within the tests. If someday in future login route malfunctions and various tests fails because of that, we can not be sure whether our tests are failing because of login route or problem with implementation of app. I am not sure how to mock logging-in in tests.

File: apps/backend/src/tests/neighborhood_api.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxHass12) [reply apps/backend/src/tests/neighborhood_api.test.ts:94]: Right now guest and non-member users are getting the same info about a neighborhood. We might want to change that.

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 7]
```diff
@@ -36,135 +37,76 @@ beforeAll(async () => {
...
-    const neighborhood = await prismaClient.neighborhood.findFirst({
+    const neighborhood: NeighborhoodDetailsForMembers = await prismaClient.neighborhood.findFirst({
       where: { name: "Bob's Neighborhood" },
       include: {
         admin: true,
         users: true,
         requests: true,
       },
-    });
+    }) as NeighborhoodDetailsForMembers;
```
Comment (nina-dk) [inline apps/backend/src/tests/neighborhood_api.test.ts:78]: I think the type assertion here is redundant. To avoid the error you're getting when removing it, you could add `const neighborhood: NeighborhoodDetailsForMembers | null`.

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 8]
```diff
@@ -36,135 +37,76 @@ beforeAll(async () => {
...
+    const neighborhood: NeighborhoodDetailsForNonMembers = await prismaClient
+      .neighborhood.findFirst({
+        where: { name: "Bob's Neighborhood" },
+        select: {
+          id: true,
+          name: true,
+          description: true,
+          location: true,
+        },
+      }) as NeighborhoodDetailsForNonMembers;
```
Comment (nina-dk) [inline apps/backend/src/tests/neighborhood_api.test.ts:107]: Same here I believe.

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 9]
```diff
@@ -178,17 +120,18 @@ describe('When neighborhoods already exist in the db', () => {
...
```
Comment (nina-dk) [inline apps/backend/src/tests/neighborhood_api.test.ts:132]: and here

File: apps/backend/src/tests/neighborhood_api.test.ts [hunk 10]
```diff
@@ -199,19 +136,201 @@ describe('When neighborhoods already exist in the db', () => {
...
+      },
+    });
+
+    const bobUserId: number | undefined = bobUser?.id;
+
+    const loginResponse = await api
+      .post('/api/login')
+      .send(BOBS_LOGIN_DATA);
+
+    const { token } = loginResponse.body;
```
[the remark this answers is not in the document]
Comment (nina-dk) [reply apps/backend/src/tests/neighborhood_api.test.ts:178]: Do you think that's not necessary?

**Review by nina-dk — APPROVED** [review]
I've clarified what you asked and added a couple more comments. Let me know what you think. Otherwise, everything looks great!

File: apps/backend/src/tests/neighborhood_api.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nina-dk) [reply apps/backend/src/tests/neighborhood_api.test.ts]: I think it's best not to have any `any` types, no other reason.

File: apps/backend/src/tests/neighborhood_api.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/tests/neighborhood_api.test.ts:178]: Yes, I don't think it adds any particular usefulness. If it were a better defined type than `string`, maybe it would of been allright.
